subsys/logging/log_backend_rtt.c

```diff
@@ -0,0 +1,220 @@
+/*
+ * Copyright (c) 2018 omSquare s.r.o.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <logging/log_backend.h>
+#include <logging/log_core.h>
+#include <logging/log_msg.h>
+#include <logging/log_output.h>
+#include <rtt/SEGGER_RTT.h>
+
+
+#if CONFIG_LOG_BACKEND_RTT_MODE_DROP
+
+#define DROP_MESSAGE "\nmessages dropped: 99 \r"
+#define DROP_MESSAGE_LEN (sizeof(DROP_MESSAGE) - 1)
+static const char *drop_msg = DROP_MESSAGE;
+static int drop_cnt = 0;
+static int drop_warn = 0;
+
+#else
+
+#define DROP_MESSAGE_LEN 0
+
+#endif /* CONFIG_LOG_BACKEND_RTT_MODE_DROP */
+
+#if CONFIG_LOG_BACKEND_RTT_BUFFER > 0
+
+#define RTT_LOCK()
+#define RTT_UNLOCK()
+static u8_t rtt_buf[CONFIG_LOG_BACKEND_RTT_BUFFER_SIZE];
+
+#else
+
+#define RTT_LOCK() SEGGER_RTT_LOCK()
+#define RTT_UNLOCK() SEGGER_RTT_UNLOCK()
+
+#endif /* CONFIG_LOG_BACKEND_RTT_BUFFER > 0 */
+
+static u8_t line_buf[CONFIG_LOG_BACKEND_RTT_MESSAGE_SIZE + DROP_MESSAGE_LEN];
+static u8_t *line_pos;
+static u8_t char_buf;
+static int panic_mode;
+
+static int msg_out(u8_t *data, size_t length, void *ctx);
+
+static int line_out(u8_t data);
+
+static void log_backend_rtt_flush(void);
+
+static int log_backend_rtt_write(void);
+
+static int log_backend_rtt_panic(u8_t *data, size_t length);
+
+static int msg_out(u8_t *data, size_t length, void *ctx)
+{
+	(void) ctx;
+	u8_t *pos;
+
+	if (panic_mode) {
+		return log_backend_rtt_panic(data, length);
+	}
+
+	for (pos = data; pos < data + length; pos++) {
+		if (line_out(*pos)) {
+			break;
+		}
+	}
+
+	return (int) (pos - data);
+}
+
+static int line_out(u8_t data)
```

Comment:
could you explain bit more complexity of writing to RTT buffer. Seems to me that you could use log_output module in a better way, by providing bigger buffer instead of 1 byte (`LOG_OUTPUT_DEFINE(log_output, msg_out, &char_buf, 1)`). Then msg_out would be called only when this buffer is full or at the end of log message with the remainder.

Reply:
Well this is all about how out_func in log_output.c behave, it will not call log_output_flush until output buffer is full. I'd like to send log message to host as soon as EOL is reached without futher delays.